File: extensions/standard-processors/processors/LogAttribute.cpp
/**
 * @file LogAttribute.cpp
 * LogAttribute class implementation
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "LogAttribute.h"
#include <cstring>
#include <memory>
#include <string>
#include <vector>
#include <queue>
#include <map>
#include <sstream>
#include <iostream>

#include "core/ProcessContext.h"
#include "core/ProcessSession.h"
#include "core/Resource.h"
#include "utils/TimeUtil.h"
#include "utils/StringUtils.h"
#include "utils/ProcessorConfigUtils.h"

namespace org::apache::nifi::minifi::processors {

void LogAttribute::initialize() {
  setSupportedProperties(Properties);
  setSupportedRelationships(Relationships);
}

void LogAttribute::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
  flowfiles_to_log_ = utils::parseU64Property(context, FlowFilesToLog);
  logger_->log_debug("FlowFiles To Log: {}", flowfiles_to_log_);

  hexencode_ = utils::parseBoolProperty(context, HexencodePayload);

  max_line_length_ = utils::parseU64Property(context, MaxPayloadLineLength);
  logger_->log_debug("Maximum Payload Line Length: {}", max_line_length_);

  if (auto attributes_to_log_str = context.getProperty(AttributesToLog)) {
    if (auto attrs_to_log_vec = utils::string::split(*attributes_to_log_str, ","); !attrs_to_log_vec.empty())
      attributes_to_log_.emplace(std::make_move_iterator(attrs_to_log_vec.begin()), std::make_move_iterator(attrs_to_log_vec.end()));
  }

  if (auto attributes_to_ignore_str = context.getProperty(AttributesToIgnore)) {
    if (auto attrs_to_ignore_vec = utils::string::split(*attributes_to_ignore_str, ","); !attrs_to_ignore_vec.empty())
      attributes_to_ignore_.emplace(std::make_move_iterator(attrs_to_ignore_vec.begin()), std::make_move_iterator(attrs_to_ignore_vec.end()));
  }

  if (auto log_level_str = context.getProperty(LogLevel)) {
    if (auto result = magic_enum::enum_cast<core::logging::LOG_LEVEL>(*log_level_str)) {
      log_level_ = *result;
    } else if (*log_level_str == "error") {  // TODO(MINIFICPP-2294) this could be avoided if config files were properly migrated
      log_level_ = core::logging::err;
    }
  }

  if (auto log_prefix = context.getProperty(LogPrefix)) {
    dash_line_ = fmt::format("{:-^50}", *log_prefix);
  }

  log_payload_ = utils::parseBoolProperty(context, LogPayload);
}

std::string LogAttribute::generateLogMessage(core::ProcessSession& session, const std::shared_ptr<core::FlowFile>& flow_file) const {
  std::ostringstream message;
  message << "Logging for flow file" << "\n";
  message << dash_line_;
  message << "\nStandard FlowFile Attributes";
  message << "\n" << "UUID:" << flow_file->getUUIDStr();
  message << "\n" << "EntryDate:" << utils::timeutils::getTimeStr(flow_file->getEntryDate());
  message << "\n" << "lineageStartDate:" << utils::timeutils::getTimeStr(flow_file->getlineageStartDate());
  message << "\n" << "Size:" << flow_file->getSize() << " Offset:" << flow_file->getOffset();
  message << "\nFlowFile Attributes Map Content";
  for (const auto& [attr_key, attr_value] : flow_file->getAttributes()) {
    if (attributes_to_ignore_ && attributes_to_ignore_->contains(attr_key))
      continue;
    if (attributes_to_log_ && !attributes_to_log_->contains(attr_key))
      continue;
    message << "\n" << "key:" << attr_key << " value:" << attr_value;
  }
  message << "\nFlowFile Resource Claim Content";
  if (const auto claim = flow_file->getResourceClaim()) {
    message << "\n" << "Content Claim:" << claim->getContentFullPath();
  }
  if (log_payload_ && flow_file->getSize() <= 1024 * 1024) {
    message << "\n" << "Payload:" << "\n";
    const auto read_result = session.readBuffer(flow_file);

    std::string printable_payload;
    if (hexencode_) {
      printable_payload = utils::string::to_hex(read_result.buffer);
    } else {
      printable_payload = to_string(read_result);
    }

    if (max_line_length_ == 0U) {
      message << printable_payload << "\n";
    } else {
      for (size_t j = 0; j < printable_payload.size(); j += max_line_length_) {
        message << printable_payload.substr(j, max_line_length_) << '\n';
      }
    }
  } else {
    message << "\n";
  }
  message << dash_line_;
  return message.str();
}

void LogAttribute::onTrigger(core::ProcessContext&, core::ProcessSession& session) {
  logger_->log_trace("enter log attribute, attempting to retrieve {} flow files", flowfiles_to_log_);
  const auto max_flow_files_to_process = flowfiles_to_log_ == 0 ? UINT64_MAX : flowfiles_to_log_;
  uint64_t flow_files_processed = 0;
  for (; flow_files_processed < max_flow_files_to_process; ++flow_files_processed) {
    std::shared_ptr<core::FlowFile> flow = session.get();

    if (!flow) {
      break;
    }

    logger_->log_with_level(log_level_, "{}", generateLogMessage(session, flow));
    session.transfer(flow, Success);
  }
  logger_->log_debug("Logged {} flow files", flow_files_processed);
}

REGISTER_RESOURCE(LogAttribute, Processor);

}  // namespace org::apache::nifi::minifi::processors

File: minifi_rust/extensions/minifi_rs_playground/src/processors/log_attribute.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

// This is the (not production ready) reimplementation of the already existing standard LogAttribute processor

use crate::processors::log_attribute::properties::{FLOW_FILES_TO_LOG, LOG_LEVEL, LOG_PAYLOAD};
use minifi_native::macros::ComponentIdentifier;
use minifi_native::{
    GetProperty, LogLevel, Logger, MinifiError, OnTriggerResult, ProcessContext, ProcessSession,
    Property, Schedule, Trigger, debug, log, trace,
};

mod properties;
mod relationships;

#[derive(Debug, ComponentIdentifier)]
pub(crate) struct LogAttributeRs {
    log_level: LogLevel,
    attributes_to_log: Option<Vec<String>>,
    attributes_to_ignore: Option<Vec<String>>,
    log_payload: bool,
    flow_files_to_log: usize,
    dash_line: String,
    hex_encode_payload: bool,
}

impl LogAttributeRs {
    fn generate_log_message<PS>(&self, session: &mut PS, flow_file: &PS::FlowFile) -> String
    where
        PS: ProcessSession,
    {
        let mut log_msg = String::with_capacity(1024);
        log_msg.push_str("Logging for flow file\n");
        log_msg.push_str(self.dash_line.as_str());

        log_msg.push_str("\nFlowFile Attributes Map Content");
        session.for_each_attribute(flow_file, |key, value| {
            if let Some(attributes_to_ignore) = &self.attributes_to_ignore
                && attributes_to_ignore.iter().any(|ign| ign == key)
            {
                return;
            }
            if let Some(attributes_to_log) = &self.attributes_to_log
                && !attributes_to_log.iter().any(|ign| ign == key)
            {
                return;
            }
            log_msg.push_str(format!("\nkey:{} value:{}", key, value).as_str());
        });
        if self.log_payload {
            log_msg.push_str("\nPayload:\n");
            if let Some(flow_file_payload) = session.read(flow_file) {
                if self.hex_encode_payload {
                    log_msg.push_str(&hex::encode(flow_file_payload));
                } else {
                    log_msg.push_str(
                        String::from_utf8(flow_file_payload)
                            .unwrap_or_default()
                            .as_str(),
                    );
                }
            }
        }
        log_msg.push('\n');
        log_msg.push_str(self.dash_line.as_str());
        log_msg
    }
}

impl Trigger for LogAttributeRs {
    fn trigger<PC, PS, L>(
        &self,
        _context: &mut PC,
        session: &mut PS,
        logger: &L,
    ) -> Result<OnTriggerResult, MinifiError>
    where
        PC: ProcessContext,
        PS: ProcessSession<FlowFile = PC::FlowFile>,
        L: Logger,
    {
        trace!(
            logger,
            "enter log attribute, attempting to retrieve {} flow files", self.flow_files_to_log
        );
        let max_flow_files_to_process = if self.flow_files_to_log == 0 {
            usize::MAX
        } else {
            self.flow_files_to_log
        };
        let mut flow_files_processed = 0usize;
        for _ in 0..max_flow_files_to_process {
            if let Some(flow_file) = session.get() {
                let log_msg = self.generate_log_message(session, &flow_file);
                log!(logger, self.log_level, "{}", log_msg);
                session.transfer(flow_file, relationships::SUCCESS.name)?;
                flow_files_processed += 1;
            } else {
                break;
            }
        }
        debug!(logger, "Logged {} flow files", flow_files_processed);

        Ok(OnTriggerResult::Ok)
    }
}

impl Schedule for LogAttributeRs {
    fn schedule<P: GetProperty, L>(context: &P, _logger: &L) -> Result<Self, MinifiError> {
        let log_level = context.get_req_property::<LogLevel>(&LOG_LEVEL)?;
        let log_payload = context.get_req_property::<bool>(&LOG_PAYLOAD)?;
        let flow_files_to_log = context.get_req_property::<usize>(&FLOW_FILES_TO_LOG)?;

        fn get_csv_property<P: GetProperty>(
            context: &P,
            property: &Property,
        ) -> Result<Option<Vec<String>>, MinifiError> {
            Ok(context.get_property::<String>(property)?.map(|s| {
                s.split(',')
                    .map(|s| s.trim().to_string())
                    .collect::<Vec<String>>()
            }))
        }

        let attributes_to_log = get_csv_property(context, &properties::ATTRIBUTES_TO_LOG)?;
        let attributes_to_ignore = get_csv_property(context, &properties::ATTRIBUTES_TO_IGNORE)?;

        let dash_line = format!(
            "{:-^50}",
            context
                .get_property::<String>(&properties::LOG_PREFIX)?
                .unwrap_or_default()
        );

        let hex_encode_payload =
            context.get_req_property::<bool>(&properties::HEX_ENCODE_PAYLOAD)?;

        Ok(LogAttributeRs {
            log_level,
            attributes_to_log,
            attributes_to_ignore,
            log_payload,
            flow_files_to_log,
            dash_line,
            hex_encode_payload,
        })
    }
}

pub(crate) mod processor_definition;

#[cfg(test)]
mod tests;
